### ISSUE TITLE
Complex values of PCA

### ISSUE THREAD

**Issue by zxdawn** [spectralpython/spectral#155]
I have tested `principal_components` for a kernel matrix and got complex numbers.

Here's an example. The small data array is uploaded here: [test.npy.zip](https://github.com/spectralpython/spectral/files/12514324/test.npy.zip).

```
import numpy as np
from spectral.algorithms import principal_components
from sklearn.metrics.pairwise import polynomial_kernel


data = np.load('test.npy')

kernel_matrix = polynomial_kernel(data, data, degree=3)

pc = principal_components(kernel_matrix)

print('eigenvectors', pc.eigenvectors)
print('eigenvalues', pc.eigenvalues)
```

Output:
```
eigenvectors [[-0.03740612+0.j          0.02968214+0.j          0.03906663+0.j
  ... -0.0014373 +0.01769771j  0.00130714-0.00076817j
   0.00130714+0.00076817j]
 [-0.02771749+0.j          0.01681404+0.j          0.03382825+0.j
  ... -0.00218317-0.0126778j  -0.0026823 -0.0009597j
  -0.0026823 +0.0009597j ]
 [-0.02419515+0.j          0.01263799+0.j          0.04935432+0.j
  ...  0.01256403-0.00298363j -0.00827857-0.00591223j
  -0.00827857+0.00591223j]
 ...
 [-0.01594176+0.j          0.00813022+0.j         -0.0241535 +0.j
  ... -0.04218511-0.01735605j -0.0269736 +0.00105019j
  -0.0269736 -0.00105019j]
 [-0.01690873+0.j          0.00890055+0.j         -0.02702997+0.j
  ...  0.03493841+0.03010962j  0.00976974+0.00470787j
   0.00976974-0.00470787j]
 [-0.01679133+0.j          0.00858658+0.j         -0.01948129+0.j
  ...  0.00937295+0.00904927j  0.01989747+0.00284286j
   0.01989747-0.00284286j]]
eigenvalues [ 6.35225437e-01+0.00000000e+00j  4.41713743e-04+0.00000000e+00j
  1.61319973e-05+0.00000000e+00j ... -1.43517616e-18-9.31143950e-19j
 -1.83147967e-18+1.66234509e-19j -1.83147967e-18-1.66234509e-19j]
```

I see the complex part is small, is it reasonable to get the real parts directly?
```
print('eigenvectors', pc.eigenvectors.real)
print('eigenvalues', pc.eigenvalues.real)

eigenvectors [[-0.03740612  0.02968214  0.03906663 ... -0.0014373   0.00130714
   0.00130714]
 [-0.02771749  0.01681404  0.03382825 ... -0.00218317 -0.0026823
  -0.0026823 ]
 [-0.02419515  0.01263799  0.04935432 ...  0.01256403 -0.00827857
  -0.00827857]
 ...
 [-0.01594176  0.00813022 -0.0241535  ... -0.04218511 -0.0269736
  -0.0269736 ]
 [-0.01690873  0.00890055 -0.02702997 ...  0.03493841  0.00976974
   0.00976974]
 [-0.01679133  0.00858658 -0.01948129 ...  0.00937295  0.01989747
   0.01989747]]
eigenvalues [ 6.35225437e-01  4.41713743e-04  1.61319973e-05 ... -1.43517616e-18
 -1.83147967e-18 -1.83147967e-18]
```

**Comment by tboggs** [comment]
I don't think you can simply ignore the complex part. The problem appears to be that the rank of your matrix is too low.

Let's start with the original data:

```
In [4]: data = np.load('test.npy')

In [5]: data.shape
Out[5]: (1280, 48)

In [6]: np.linalg.matrix_rank(data)
Out[6]: 48
```

You have 1280 vectors with 48 features/dimensions and the rank of the data matrix is equal to the number of features. So it should be fine to do PCA:

```
In [7]: spy.principal_components(data).eigenvalues.dtype
Out[7]: dtype('float64')
```
Now, let's look at your kernel matrix.

```
In [8]: kernel_matrix = polynomial_kernel(data, data, degree=3)

In [9]: kernel_matrix.shape
Out[9]: (1280, 1280)

In [10]: np.linalg.matrix_rank(kernel_matrix)
Out[10]: 3
```

You've transformed the original 48 features into 1280 features. But since your new kernel matrix has a rank of only 3 (which is much less than 1280), the covariance matrix will be singular so inverting it will be unstable, resulting in complex eigenvalues.

```
In [11]: spy.principal_components(kernel_matrix).eigenvalues.dtype
Out[11]: dtype('complex128')
```

In short, you should ensure that your matrix rank is equal to the number of features before attempting PCA.

**Comment by zxdawn** [comment]
Thank you for your explanation @tboggs! I understand the issue now.

Actually, I'm trying to apply kernel PCA as mentioned in the appendix of [this paper](https://doi.org/10.1007/s11263-006-6689-3):

> K = K(X, X) = (K)ij the N × N kernel matrix whose entries are the dot products 〈φ(xi ),φ(x j )〉. φ is a nonlinear mapping associated with a kernel function.

It seems I misunderstood the equation and `polynomial_kernel(data, data, degree=3)` is wrong ...

Then, I found this [tutorial](https://medium.com/@phsamuel.work/kernel-pca-f7a40f0b4265) of kernel PCA and tried to apply it:

```
import numpy as np

def gauss_kernel(x,y,sigma2):
    return np.exp(-(x-y).transpose().dot(x-y)/sigma2)

def gram(x,y,kernel):
    g=np.zeros((x.shape[1],y.shape[1]))
    for i in range(x.shape[1]):
        for j in range(y.shape[1]):
            g[i,j]=kernel(x[:,i],y[:,j])
    return g

data = np.load('test.npy')
G=gram(data,data, lambda x,y: gauss_kernel(x,y,4.))

pc = principal_components(G)

print('eigenvectors', pc.eigenvectors.real)
print('eigenvalues', pc.eigenvalues.real)
```

It works well! So, what's the difference between replacing the kernel with polynomial manually and 
 using `polynomial_kernel(data, data, degree=3)` directly?

**Comment by zxdawn** [comment]
I finally figured out how to calculate the kernel PCA and compared my method with sklearn [here](https://github.com/scikit-learn/scikit-learn/discussions/27298). Since it's not the issue of spectral, I close it now. If you have any idea of the kernel PCA differences, please feel free to comment :)

BTW, maybe adding kernel PCA to spectral is also a good option in the future.

**Comment by tboggs** [comment]
I'll take a look at adding it as an option to the `principal_components` function. And if I get some free time, I'll take a look at the differences between your version and sklearn.
